### PR TITLE
feat: Migrate to TensorFlow v2.0

### LINKED ISSUE #539: Migrate to TensorFlow 2.0
# Description

The API in [TensorFlow 2.0](https://github.com/tensorflow/community/blob/master/sigs/testing/faq.md) changes quite a bit, and the concept of sessions goes away. There is a [migration guide here](https://www.tensorflow.org/beta/guide/migration_guide). The most work that will need to be done is just in the optimizer, but as an example of how things will be different in the documentation see the following example for `pyhf.tensorlib.clip` in TF v1.14 code

```python
>>> import pyhf
>>> import tensorflow as tf
>>> sess = tf.Session()
...
>>> pyhf.set_backend(pyhf.tensor.tensorflow_backend(session=sess))
>>> a = pyhf.tensorlib.astensor([-2, -1, 0, 1, 2])
>>> with sess.as_default():
...   sess.run(pyhf.tensorlib.clip(a, -1, 1))
...
array([-1., -1.,  0.,  1.,  1.], dtype=float32)
```

vs. TF 2.0 code

```python
>>> import pyhf
>>> pyhf.set_backend(pyhf.tensor.tensorflow_backend())
>>> a = pyhf.tensorlib.astensor([-2, -1, 0, 1, 2])
>>> pyhf.tensorlib.clip(a, -1, 1)
<tf.Tensor: id=5, shape=(5,), dtype=float32, numpy=array([-1., -1.,  0.,  1.,  1.], dtype=float32)>
>>> pyhf.tensorlib.clip(a, -1, 1).numpy()
array([-1., -1.,  0.,  1.,  1.], dtype=float32)
```

or for later evaluation with other actions

```python
>>> import pyhf
>>> import tensorflow as tf
>>> pyhf.set_backend(pyhf.tensor.tensorflow_backend())
>>> a = pyhf.tensorlib.astensor([-2, -1, 0, 1, 2])
>>> g = tf.function(pyhf.tensorlib.clip) # g for graph, btw
>>> g(a, -1, 1)
<tf.Tensor: id=82, shape=(5,), dtype=float32, numpy=array([-1., -1.,  0.,  1.,  1.], dtype=float32)>
>>> g(a, -1, 1).numpy()
array([-1., -1.,  0.,  1.,  1.], dtype=float32)
```

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #539 

Instead of adding conditional control flows through PR #535 and PR #537 in support of PR #534 , PR #534 will be delayed in favor of migrating to TensorFlow v2.0, where eager evaluation will avoid the issues that PR #535 and PR #537 addressed.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Migrate the TensorFlow backend and optimizer to TensorFlow v2.0 / deprecate TensorFlow v1.0 behavior
   - Changes evaluation from lazy (v1.0) to eager (v2.0 default)
* Require TensorFlow and TensorFlow Probability releases compatible with TensorFlow v2.0
* Simplify test configuration with deprecation of TensorFlow v1.0 Sessions
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
**TODO:**
- [ ] Rewrite `opt_tflow.py` using [`tf.function`](https://www.tensorflow.org/beta/guide/autograph) instead of [`tf.compat.v1.Session`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/compat/v1/Session#run)
   - This is a bit confusing for me at the moment to be honest.

References:
- [RFC TF 2.0: Functions, not Sessions](https://github.com/tensorflow/community/blob/master/rfcs/20180918-functions-not-sessions-20.md)

**Comment by matthewfeickert** [comment]
Until `tf.hessians` gets updated in TF 2.0 this PR is [blocked as a results of gradients](https://github.com/tensorflow/tensorflow/issues/29781). So this will need to wait for a new release candidate.

**Comment by matthewfeickert** [comment]
Given the decision to [revert the addition of "improvement" in `commitizen/conventional-commit-types`](https://github.com/commitizen/conventional-commit-types/pull/13#issuecomment-549189892) this was renamed too a feature.

**Comment by matthewfeickert** [comment]
The coverage needs to get fixed here

https://codecov.io/gh/scikit-hep/pyhf/pull/541/diff?src=pr&el=tree#D3-102

but other than that this looks like it is almost ready for review.

**Comment by lukasheinrich** [comment]
this looks great! thanks you @matthewfeickert. After this transition it might be worth looking at performance. I'm not sure we understand yet why TF is so slow (maybe one can optimizer similar to JAX), but this PR will set us up nicely for the future.

**Comment by lukasheinrich** [comment]
just to understand better: this does not enable eager execution, but just removes all the session stuff and handles gradients differently?

**Comment by matthewfeickert** [comment]
> just to understand better: this does not enable eager execution, but just removes all the session stuff and handles gradients differently?

No this _does_ enable eager execution. If you check out the docstrings that I added that's what we'll be getting from now on.

**Comment by lukasheinrich** [comment]
ah right, sorry I overlooked this

**Comment by matthewfeickert** [comment]
> ah right, sorry I overlooked this

No I should have mentioned this in the PR body. I've now updated the commit history message for it.

File: setup.py [hunk 1]
```diff
@@ -9,7 +9,11 @@
     long_description = readme_md.read()
 
 extras_require = {
-    'tensorflow': ['tensorflow~=1.15', 'tensorflow-probability~=0.8', 'numpy~=1.16'],
+    'tensorflow': [
+        'tensorflow~=2.0',
+        'tensorflow-probability~=0.8',
+        'numpy<2.0,>=1.16.0',
```
Comment (kratsg) [inline setup.py]: do we need to keep this around given: https://github.com/tensorflow/tensorflow/blob/5364121e858be715439c328c5f67de8441050f4b/tensorflow/tools/pip_package/setup.py#L61

Comment (matthewfeickert) [reply setup.py]: Yeah, we can simplify that to just `numpy~=1.16`. Good point!

**Pull request merged**